File: processing/feature_views.py
"""Pure feature extraction helpers for Phy-inspired spike feature views."""

import re

import numpy as np

from processing.cluster_diagnostics import normalize_cluster_ids


def _finite_number(value):
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if np.isfinite(number) else None


def _spike_channel(spike):
    value = _finite_number(spike.get('channel')) if isinstance(spike, dict) else None
    return int(round(value)) if value is not None else None


def _feature_values(spike, sample_rate_hz):
    values = {}
    time_samples = _finite_number(spike.get('time'))
    if time_samples is not None:
        values['timeSeconds'] = time_samples / sample_rate_hz
    pc1 = _finite_number(spike.get('x'))
    pc2 = _finite_number(spike.get('y'))
    if pc1 is not None:
        values['pc1'] = pc1
    if pc2 is not None:
        values['pc2'] = pc2
    channel = _spike_channel(spike)
    if channel is not None:
        values['channel'] = channel

    for key in ('amplitude', 'amp', 'peakAmplitude', 'spikeAmplitude'):
        amplitude = _finite_number(spike.get(key))
        if amplitude is not None:
            values['amplitude'] = amplitude
            break

    for key, value in spike.items():
        match = re.fullmatch(r'pc[_-]?(\d+)', str(key), re.IGNORECASE)
        number = _finite_number(value)
        if match and number is not None:
            values[f'pc{int(match.group(1))}'] = number

    features = spike.get('features')
    if isinstance(features, dict):
        for key, value in features.items():
            number = _finite_number(value)
            if number is not None:
                safe_key = re.sub(r'[^a-zA-Z0-9_]+', '_', str(key)).strip('_') or 'value'
                values[f'feature_{safe_key}'] = number
    elif isinstance(features, (list, tuple, np.ndarray)):
        for index, value in enumerate(features):
            number = _finite_number(value)
            if number is not None:
                values[f'feature{index + 1}'] = number

    template_features = spike.get('template_features', spike.get('templateFeatures'))
    if isinstance(template_features, (list, tuple, np.ndarray)):
        for index, value in enumerate(template_features):
            number = _finite_number(value)
            if number is not None:
                values[f'templateFeature{index + 1}'] = number
    else:
        number = _finite_number(spike.get('templateFeature'))
        if number is not None:
            values['templateFeature1'] = number
    return values


def _sample_indexed_spikes(indexed, maximum):
    indexed = list(indexed or [])
    indexed.sort(key=lambda item: _finite_number(item[1].get('time')) or 0.0)
    if len(indexed) <= maximum:
        return indexed
    indices = np.linspace(0, len(indexed) - 1, num=maximum, dtype=np.int64)
    return [indexed[index] for index in indices]


def _build_point(cluster_id, point_index, spike, sample_rate_hz):
    raw_spike_index = spike.get('spikeIndex', point_index)
    try:
        spike_index = int(raw_spike_index)
    except (TypeError, ValueError):
        spike_index = int(point_index)
    time_samples = _finite_number(spike.get('time'))
    return {
        'spikeId': f'{cluster_id}:{spike_index}',
        'clusterId': cluster_id,
        'pointIndex': int(point_index),
        'spikeIndex': spike_index,
        'timeSamples': time_samples,
        'channel': _spike_channel(spike),
        'values': _feature_values(spike, sample_rate_hz),
    }


def _add_pair_projection(series):
    if len(series) != 2:
        return None
    retained_dimension = next((
        dimension
        for dimension in ('templateFeature1', 'templateFeature2')
        if all(any(dimension in point['values'] for point in item['points']) for item in series)
    ), None)
    if retained_dimension:
        for item in series:
            for point in item['points']:
                if retained_dimension in point['values']:
                    point['values']['pairProjection'] = point['values'][retained_dimension]
        return 'retained_template_features'

    centroid_values = []
    for item in series:
        coordinates = [
            (point['values']['pc1'], point['values']['pc2'])
            for point in item['points']
            if 'pc1' in point['values'] and 'pc2' in point['values']
        ]
        if not coordinates:
            return None
        centroid_values.append(np.mean(np.asarray(coordinates, dtype=np.float64), axis=0))
    axis = centroid_values[1] - centroid_values[0]
    norm = np.linalg.norm(axis)
    axis = axis / norm if norm > 0 else np.asarray([1.0, 0.0])
    origin = (centroid_values[0] + centroid_values[1]) / 2.0
    for item in series:
        for point in item['points']:
            if 'pc1' not in point['values'] or 'pc2' not in point['values']:
                continue
            coordinates = np.asarray([point['values']['pc1'], point['values']['pc2']])
            point['values']['pairProjection'] = float(np.dot(coordinates - origin, axis))
    return 'pca_centroid_axis'


def _dimension_metadata(dimension_id):
    labels = {
        'timeSeconds': ('Recording time', 's', 'spike_time'),
        'pc1': ('PC 1', None, 'retained_embedding'),
        'pc2': ('PC 2', None, 'retained_embedding'),
        'channel': ('Channel', 'channel_id', 'spike_assignment'),
        'amplitude': ('Amplitude', None, 'sorter_or_spike_metadata'),
        'pairProjection': ('Pair separation', None, 'pair_projection'),
    }
    label, unit, source = labels.get(
        dimension_id,
        (re.sub(r'(?<!^)(?=[A-Z])', ' ', dimension_id).replace('_', ' ').title(), None, 'spike_metadata'),
    )
    return {'id': dimension_id, 'label': label, 'unit': unit, 'source': source}


def extract_cluster_features(
    clustering_results,
    cluster_ids,
    sample_rate_hz=30000.0,
    max_spikes_per_cluster=5000,
    include_background=True,
    max_background_spikes=5000,
    selected_channels=None,
):
    """Return bounded, stable per-spike features and optional gray background."""
    sample_rate_hz = max(float(sample_rate_hz), 1.0)
    cluster_ids = normalize_cluster_ids(clustering_results, cluster_ids, limit=12)
    maximum = min(max(int(max_spikes_per_cluster), 10), 20000)
    background_maximum = min(max(int(max_background_spikes), 0), 20000)
    channel_filter = {
        int(round(float(channel)))
        for channel in (selected_channels or [])
        if _finite_number(channel) is not None
    }

    series = []
    for cluster_id in cluster_ids:
        indexed_spikes = [
            (point_index, spike)
            for point_index, spike in enumerate(clustering_results[cluster_id] or [])
            if not channel_filter or _spike_channel(spike) in channel_filter
        ]
        points = [
            _build_point(cluster_id, point_index, spike, sample_rate_hz)
            for point_index, spike in _sample_indexed_spikes(
                indexed_spikes,
                maximum,
            )
        ]
        series.append({
            'clusterId': cluster_id,
            'points': points,
            'totalSpikes': len(clustering_results[cluster_id] or []),
            'returnedSpikes': len(points),
        })

    pair_feature_source = _add_pair_projection(series)
    background_points = []
    if include_background and background_maximum:
        selected = set(cluster_ids)
        candidates = [
            (cluster_id, point_index, spike)
            for cluster_id, spikes in enumerate(clustering_results or [])
            if cluster_id not in selected
            for point_index, spike in enumerate(spikes or [])
            if not channel_filter or _spike_channel(spike) in channel_filter
        ]
        if len(candidates) > background_maximum:
            indices = np.linspace(0, len(candidates) - 1, num=background_maximum, dtype=np.int64)
            candidates = [candidates[index] for index in indices]
        background_points = [
            _build_point(cluster_id, point_index, spike, sample_rate_hz)
            for cluster_id, point_index, spike in candidates
        ]

    dimension_ids = set()
    for item in series:
        for point in item['points']:
            dimension_ids.update(point['values'])
    for point in background_points:
        dimension_ids.update(point['values'])
    preferred = ['timeSeconds', 'pc1', 'pc2', 'pairProjection', 'amplitude', 'channel']
    ordered_dimensions = [dimension for dimension in preferred if dimension in dimension_ids]
    ordered_dimensions.extend(sorted(dimension_ids - set(ordered_dimensions)))

    return {
        'clusterIds': cluster_ids,
        'sampleRateHz': sample_rate_hz,
        'dimensions': [_dimension_metadata(dimension) for dimension in ordered_dimensions],
        'series': series,
        'backgroundPoints': background_points,
        'pairFeatureSource': pair_feature_source,
        'selectedChannels': sorted(channel_filter),
    }

File: src/utils/featureViews.js
const finiteNumber = (value) => {
  const number = Number(value);
  return Number.isFinite(number) ? number : null;
};

const clusterIdOf = (cluster, fallback) => cluster?.clusterId ?? cluster?.id ?? fallback;

const dimensionMetadata = (id) => {
  const definitions = {
    timeSeconds: ['Recording time', 's', 'spike_time'],
    pc1: ['PC 1', null, 'retained_embedding'],
    pc2: ['PC 2', null, 'retained_embedding'],
    channel: ['Channel', 'channel_id', 'spike_assignment'],
    amplitude: ['Amplitude', null, 'sorter_or_spike_metadata'],
    pairProjection: ['Pair separation', null, 'pair_projection'],
  };
  const [label, unit, source] = definitions[id] || [
    id.replace(/([A-Z])/g, ' $1').replace(/_/g, ' ').trim(),
    null,
    'spike_metadata',
  ];
  return { id, label, unit, source };
};

const valuesFromSpike = (spike, sampleRateHz) => {
  const values = {};
  const timeSamples = finiteNumber(spike?.time ?? spike?.timeSamples);
  const pc1 = finiteNumber(spike?.x ?? spike?.pc1);
  const pc2 = finiteNumber(spike?.y ?? spike?.pc2);
  const channel = finiteNumber(spike?.channel);
  const amplitude = finiteNumber(
    spike?.amplitude ?? spike?.amp ?? spike?.peakAmplitude ?? spike?.spikeAmplitude
  );
  if (timeSamples !== null) values.timeSeconds = timeSamples / sampleRateHz;
  if (pc1 !== null) values.pc1 = pc1;
  if (pc2 !== null) values.pc2 = pc2;
  if (channel !== null) values.channel = Math.round(channel);
  if (amplitude !== null) values.amplitude = amplitude;

  const featureValues = spike?.features;
  if (Array.isArray(featureValues)) {
    featureValues.forEach((value, index) => {
      const number = finiteNumber(value);
      if (number !== null) values[`feature${index + 1}`] = number;
    });
  } else if (featureValues && typeof featureValues === 'object') {
    Object.entries(featureValues).forEach(([key, value]) => {
      const number = finiteNumber(value);
      const safeKey = String(key).replace(/[^a-zA-Z0-9_]+/g, '_').replace(/^_+|_+$/g, '') || 'value';
      if (number !== null) values[`feature_${safeKey}`] = number;
    });
  }

  const templateFeatures = spike?.template_features ?? spike?.templateFeatures;
  if (Array.isArray(templateFeatures)) {
    templateFeatures.forEach((value, index) => {
      const number = finiteNumber(value);
      if (number !== null) values[`templateFeature${index + 1}`] = number;
    });
  } else {
    const templateFeature = finiteNumber(spike?.templateFeature);
    if (templateFeature !== null) values.templateFeature1 = templateFeature;
  }
  return values;
};

const clusterEntries = ({ clusterData, clusteringResults }) => {
  if (Array.isArray(clusteringResults?.fullData) && clusteringResults.fullData.length) {
    return clusteringResults.fullData.map((spikes, index) => ({
      clusterId: clusterIdOf(clusteringResults.clusters?.[index], index),
      spikes: (spikes || []).map((spike, pointIndex) => ({ ...spike, pointIndex })),
    }));
  }

  return (clusterData?.clusters || []).map((cluster, index) => {
    const points = cluster.points || [];
    const times = cluster.spikeTimes || cluster.spike_times || [];
    const channels = cluster.spikeChannels || [];
    const amplitudes = cluster.spikeAmplitudes || [];
    const spikeIndices = cluster.spikeIndices || [];
    const count = Math.max(points.length, times.length, channels.length, amplitudes.length);
    return {
      clusterId: clusterIdOf(cluster, index),
      spikes: Array.from({ length: count }, (_, pointIndex) => ({
        x: points[pointIndex]?.[0],
        y: points[pointIndex]?.[1],
        time: times[pointIndex],
        channel: channels[pointIndex] ?? cluster.channelId,
        amplitude: amplitudes[pointIndex],
        spikeIndex: spikeIndices[pointIndex] ?? pointIndex,
        pointIndex,
      })),
    };
  });
};

const evenlySample = (items, maximum) => {
  if (items.length <= maximum) return items;
  if (maximum <= 1) return items.slice(0, maximum);
  return Array.from({ length: maximum }, (_, index) => (
    items[Math.round(index * (items.length - 1) / (maximum - 1))]
  ));
};

const buildPoint = (clusterId, spike, fallbackIndex, sampleRateHz) => {
  const pointIndex = finiteNumber(spike?.pointIndex) ?? fallbackIndex;
  const spikeIndex = finiteNumber(spike?.spikeIndex) ?? pointIndex;
  const timeSamples = finiteNumber(spike?.time ?? spike?.timeSamples);
  const channel = finiteNumber(spike?.channel);
  return {
    spikeId: `${clusterId}:${Math.round(spikeIndex)}`,
    clusterId,
    pointIndex: Math.round(pointIndex),
    spikeIndex: Math.round(spikeIndex),
    timeSamples,
    channel: channel === null ? null : Math.round(channel),
    values: valuesFromSpike(spike, sampleRateHz),
  };
};

const addPairProjection = (series) => {
  if (series.length !== 2 || series.some((item) => item.points.length === 0)) return null;
  const retainedDimension = ['templateFeature1', 'templateFeature2'].find((dimension) => (
    series.every((item) => item.points.some((point) => point.values[dimension] !== undefined))
  ));
  if (retainedDimension) {
    series.forEach((item) => item.points.forEach((point) => {
      if (point.values[retainedDimension] !== undefined) {
        point.values.pairProjection = point.values[retainedDimension];
      }
    }));
    return 'retained_template_features';
  }

  const coordinates = series.map((item) => item.points
    .filter((point) => Number.isFinite(point.values.pc1) && Number.isFinite(point.values.pc2))
    .map((point) => [point.values.pc1, point.values.pc2]));
  if (coordinates.some((points) => points.length === 0)) return null;
  const centroids = coordinates.map((points) => points.reduce(
    (sum, point) => [sum[0] + point[0] / points.length, sum[1] + point[1] / points.length],
    [0, 0]
  ));
  let axis = [centroids[1][0] - centroids[0][0], centroids[1][1] - centroids[0][1]];
  const norm = Math.hypot(...axis);
  axis = norm > 0 ? axis.map((value) => value / norm) : [1, 0];
  const origin = [
    (centroids[0][0] + centroids[1][0]) / 2,
    (centroids[0][1] + centroids[1][1]) / 2,
  ];
  series.forEach((item) => item.points.forEach((point) => {
    if (!Number.isFinite(point.values.pc1) || !Number.isFinite(point.values.pc2)) return;
    point.values.pairProjection = (
      (point.values.pc1 - origin[0]) * axis[0]
      + (point.values.pc2 - origin[1]) * axis[1]
    );
  }));
  return 'pca_centroid_axis';
};

export const buildLocalFeaturePayload = ({
  clusterData = null,
  clusteringResults = null,
  clusterIds = [],
  sampleRateHz = 30000,
  maxSpikesPerCluster = 5000,
  includeBackground = true,
  maxBackgroundSpikes = 5000,
  selectedChannels = [],
} = {}) => {
  const actualSampleRate = Math.max(finiteNumber(sampleRateHz) ?? 30000, 1);
  const selectedKeys = new Set(clusterIds.map(String));
  const channelKeys = new Set(selectedChannels.map((channel) => String(Math.round(Number(channel)))));
  const entries = clusterEntries({ clusterData, clusteringResults });
  const allowsChannel = (spike) => channelKeys.size === 0 || channelKeys.has(String(Math.round(Number(spike.channel))));
  const makePoints = (entry, maximum) => evenlySample(
    entry.spikes.filter(allowsChannel),
    Math.max(0, maximum)
  ).map((spike, index) => buildPoint(entry.clusterId, spike, index, actualSampleRate));

  const series = clusterIds.map((clusterId) => {
    const entry = entries.find((candidate) => String(candidate.clusterId) === String(clusterId))
      || { clusterId, spikes: [] };
    const points = makePoints(entry, maxSpikesPerCluster);
    return {
      clusterId,
      points,
      totalSpikes: entry.spikes.length,
      returnedSpikes: points.length,
    };
  });
  const pairFeatureSource = addPairProjection(series);
  const backgroundCandidates = entries
    .filter((entry) => !selectedKeys.has(String(entry.clusterId)))
    .flatMap((entry) => entry.spikes.filter(allowsChannel).map((spike, index) => (
      buildPoint(entry.clusterId, spike, index, actualSampleRate)
    )));
  const backgroundPoints = includeBackground
    ? evenlySample(backgroundCandidates, Math.max(0, maxBackgroundSpikes))
    : [];

  const dimensionIds = new Set();
  [...series.flatMap((item) => item.points), ...backgroundPoints]
    .forEach((point) => Object.keys(point.values).forEach((id) => dimensionIds.add(id)));
  const preferred = ['timeSeconds', 'pc1', 'pc2', 'pairProjection', 'amplitude', 'channel'];
  const ordered = preferred.filter((id) => dimensionIds.has(id));
  [...dimensionIds].sort().forEach((id) => { if (!ordered.includes(id)) ordered.push(id); });

  return {
    clusterIds,
    sampleRateHz: actualSampleRate,
    dimensions: ordered.map(dimensionMetadata),
    series,
    backgroundPoints,
    pairFeatureSource,
    selectedChannels: [...channelKeys].map(Number).sort((left, right) => left - right),
  };
};

export const getFeatureSelectionFromPlotEvent = (plotEvent) => {
  const points = (plotEvent?.points || [])
    .map((point) => point.customdata)
    .filter((point) => point?.spikeId);
  return [...new Map(points.map((point) => [point.spikeId, point])).values()];
};
